RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+      //      chi2min[ipass] = 100000.f;
+      //      chi2 = 10000.f;
+      chi2 = chi2min[ipass];
+      while(chi2 <= chi2min[ipass] && niter < 15 && (niter < 2 || (fabs(xstep) > 0.2 && fabs(ystep) > 0.2))) {
+         
+         // Remember the present parameters
+         x2D0[ipass] = x2D;
+         y2D0[ipass] = y2D;
+         qtfrac0[ipass] = qtfrac;
+         xerr2[ipass] = minv11;
+         yerr2[ipass] = minv22;
+         chi2min[ipass] = chi2;
+         niter0[ipass] = niter;
+         
+         // Calculate the initial template which also allows the error calculation for the struck pixels
+         
+         for(j=0; j<BXM2; ++j) {for(i=0; i<BYM2; ++i) {template2d[j][i] = 0.f;}}
+         templ2D.xytemp(x2D, y2D, ydouble, xdouble, template2d, true, dpdx2d, qtemplate);
+         
+         float sumptdt1 = 0.f, sumptdt2 = 0.f;
+         float sumdtdt11 = 0.f, sumdtdt12 = 0.f, sumdtdt22 = 0.f;
+         chi2 = 0.f;
+         qactive = 0.f;
+         // Loop over all pixels and calculate matrices
+         
+         for(k=0; k<tpixel; ++k) {
+            int jpix = indexxy[0][k];
+            int ipix = indexxy[1][k];
+            float qtpixel = template2d[jpix][ipix];
+            float pt = pixel[k]-qtpixel;
+            chi2 += pt*pt/sigma2[k];
+            sumptdt1 += pt*dpdx2d[0][jpix][ipix]/sigma2[k];
+            sumptdt2 += pt*dpdx2d[1][jpix][ipix]/sigma2[k];
+            sumdtdt11 += dpdx2d[0][jpix][ipix]*dpdx2d[0][jpix][ipix]/sigma2[k];
+            sumdtdt12 += dpdx2d[0][jpix][ipix]*dpdx2d[1][jpix][ipix]/sigma2[k];
+            sumdtdt22 += dpdx2d[1][jpix][ipix]*dpdx2d[1][jpix][ipix]/sigma2[k];
+            if(k < npixel) {qactive += qtpixel;}
+         }
+         
+         // invert the parameter covariance matrix
+         
+         float D = sumdtdt11*sumdtdt22 - sumdtdt12*sumdtdt12;
+         minv11 = sumdtdt22/D;
+         minv12 = -sumdtdt12/D;
+         minv22 = sumdtdt11/D;
+         
+         qtfrac = qactive/qtemplate;
+         
+         //Calculate the step size
+         
+         xstep = minv11*sumptdt1 + minv12*sumptdt2;
+         xstep *= 0.9f;
+         ystep = minv12*sumptdt1 + minv22*sumptdt2;
+         ystep *= 0.9f;
+//         printf("niter = %d, chisq = %f, xstep = %f, ystep = %f \n", niter, chi2, xstep, ystep);
+         if(fabs(xstep) > 2.*xsize || fabs(ystep) > 2.*ysize) break;
+         x2D += xstep;
+         y2D += ystep;
+         ++niter;
+      }
+   }
+   
+   ipass = 0;
+   if(npass == 1) {
+      // one pass, require that it have iterated
+      if(niter0[0] == 0) {return 2;}
+   } else {
+      // two passes
+      if(niter0[0] == 0 && niter0[1] == 0) {return 2;}
+      if(niter0[0] > 0 && niter0[1] > 0) {
+         // if both have iterated, take the smaller chi2
+         if(chi2min[1] < chi2min[0]) {ipass = 1;}
+      } else {
+         // if one has iterated, take it
+         if(niter0[1] > 0) {ipass = 1;}
+      }
+   }
+   
+   //   printf("niter = %d \n", niter);

```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.